scipy/sparse/_sputils.py

```diff
@@ -150,6 +150,94 @@ def getdata(obj, dtype=None, copy=False) -> np.ndarray:
     return data
 
 
+def safely_cast_index_arrays(A, idx_dtype=np.int32, msg=""):
+    """Return cast index arrays of `A` to idx_dtype safely.
+
+    Check the shape of `A` to determine if it is safe to cast its index
+    arrays to dtype `idx_dtype`. If any dimension in shape is larger than
+    fits in the dtype, casting is unsafe so raise ``ValueError(msg)``.
+    If safe, cast the index arrays to `idx_dtype` and return the result
+    without changing the input `A`. The caller can assign results to `A`
+    attributes if desired or use the recast index arrays directly.
+
+    If downcasting is not needed, no copy is made.
+    You can test e.g. ``A.indptr is indptr`` to see if downcasting occurred.
+
+    .. versionadded:: 1.15.0
+
+    Parameters
+    ----------
+    A : sparse array or matrix
+        The array for which index arrays should be downcast.
+    idx_dtype : the index dtype desired to downcast to.
+        Should be an integer dtype. default: np.int32
+    msg : string
+        A string to be added to the end of the ValueError message
+        if the array shape is too big to fit in `idx_dtype`.
+        It should indicate why the downcasting is needed and
+        defaults to f"dtype {idx_dtype}".
+
+    Returns
+    -------
+    idx_arrays : ndarray or tuple of ndarrays
+        Based on ``A.format``, index arrays are returned after casting to idx_dtype.
+        For CSC/CSR, returns ``(indices, indptr)``.
+        For COO, returns ``coords``.
+        For DIA, returns ``offsets``.
+        For BSR, returns ``(indices, indtr)``.
+
+    Raises
+    ------
+    ValueError : if the array has shape that would not fit in the new dtype.
+    """
+    if not msg:
+        msg = f"dtype {idx_dtype}"
+    # check for safe downcasting
+    max_value = np.iinfo(idx_dtype).max
+
+    if A.format in ("csc", "csr"):
+        # indptr[-1] is max b/c indptr always sorted
+        if A.indptr[-1] > max_value:
+            raise ValueError(f"indptr values too large for {msg}")
+
+        # check shape vs dtype
+        if max(*A.shape) > max_value:
+            if (A.indices > max_value).any():
+                raise ValueError(f"indices values too large for {msg}")
+
+        indices = A.indices.astype(idx_dtype, copy=False)
+        indptr = A.indptr.astype(idx_dtype, copy=False)
+        return indices, indptr
+
+    elif A.format == "coo":
+        if max(*A.shape) > max_value:
```

Comment:
Could we pull this out as a short-circuit test? I.e.:

```
if max(*A.shape) <= max_value:
    return A.indices
```

Then we don't have to bother with all those checks, in each block, below.

Reply:
We can't really pull this out. We still need to cast the indices and their names differ by format. Plus this `if` clause is itself a short-cut of the 2nd if-clause where we check each value, e.g. `if (A.offsets > max_value).any():`.  So while the shape-if and the ValueError can be made the same, the 2nd if and the casting code sandwich these. So I don't see a way to pull them apart.